src/OrleansServiceBus/Providers/Streams/EventHub/EventHubAdapterFactory.cs

```diff
@@ -0,0 +1,168 @@
+        private EventHubQueueMapper streamQueueMapper;
+        private IStreamFailureHandler streamFailureHandler;
+        private string[] partitionIds;
+        private ConcurrentDictionary<QueueId, EventHubAdapterReceiver> receivers;
+        private EventHubClient client;
+
+        public string Name { get { return adapterConfig.StreamProviderName; } }
+        public bool IsRewindable { get { return true; } }
+        public StreamProviderDirection Direction { get { return StreamProviderDirection.ReadWrite; } }
+
+        /// <summary>
+        /// Factory initialization.
+        /// Provider config must contain the event hub settings type or the settings themselves.
+        /// EventHubSettingsType is recommended for consumers that do not want to include secure information in the cluster configuration.
+        /// </summary>
+        /// <param name="providerConfig"></param>
+        /// <param name="providerName"></param>
+        /// <param name="log"></param>
+        /// <param name="svcProvider"></param>
+        public void Init(IProviderConfiguration providerConfig, string providerName, Logger log, IServiceProvider svcProvider)
+        {
+            if (providerConfig == null) throw new ArgumentNullException("providerConfig");
+            if (string.IsNullOrWhiteSpace(providerName)) throw new ArgumentNullException("providerName");
+            if (log == null) throw new ArgumentNullException("log");
+            if (svcProvider == null) throw new ArgumentNullException("svcProvider");
+
+            logger = log;
+            serviceProvider = svcProvider;
+            adapterConfig = new EventHubStreamProviderConfig(providerName);
+            adapterConfig.PopulateFromProviderConfig(providerConfig);
+
+            // if no event hub settings type is provided, use EventHubSettings and get populate settings from providerConfig
+            if (adapterConfig.EventHubSettingsType == null)
+            {
+                adapterConfig.EventHubSettingsType = typeof (EventHubSettings);
+            }
+
+            hubSettings = serviceProvider.GetService(adapterConfig.EventHubSettingsType) as IEventHubSettings;
+            if (hubSettings == null)
+            {
+                throw new ArgumentOutOfRangeException("providerConfig", "EventHubSettingsType not valid.");
+            }
+
+            // if settings is an EventHubSettings class, populate settings from providerConfig
+            var settings = hubSettings as EventHubSettings;
+            if (settings != null)
+            {
+                settings.PopulateFromProviderConfig(providerConfig);
+            }
+
+            receivers = new ConcurrentDictionary<QueueId, EventHubAdapterReceiver>();
+            client = EventHubClient.CreateFromConnectionString(hubSettings.ConnectionString, hubSettings.Path);
+        }
+
+        public async Task<IQueueAdapter> CreateAdapter()
+        {
+            partitionIds = await GetPartitionIdsAsync();
```

Comment:
if this is called many times, then the partitionIds should be cached. Not familiar with the usage patterns